File: backend/fitness/exercises/squats.js
class SquatDetector {
  constructor() {
    this.counter = 0;
    this.stage = null;
    this.min_ang = 0;
    this.min_ang_hip = 0;
    this.last_squat_time = 0;
    this.cooldown = 1000; // 1 second cooldown
    this.consecutive_frames = 0;
    this.frames_required = 3; // Reduced for more responsive detection
    this.angle_history = []; // For smoothing
    this.history_size = 5; // Number of frames to average
  }

  calculate_angle(p1, p2, p3) {
    if (!p1 || !p2 || !p3) return 0;
    
    const radians = Math.atan2(p3.y - p2.y, p3.x - p2.x) - 
                   Math.atan2(p1.y - p2.y, p1.x - p2.x);
    let angle = Math.abs(radians * 180.0 / Math.PI);
    if (angle > 180.0) {
      angle = 360 - angle;
    }
    return angle;
  }

  smooth_angle(new_angle) {
    this.angle_history.push(new_angle);
    if (this.angle_history.length > this.history_size) {
      this.angle_history.shift();
    }
    return this.angle_history.reduce((a, b) => a + b, 0) / this.angle_history.length;
  }

  validate_form(landmarks) {
    // Get key points
    const left_hip = landmarks[23];
    const left_knee = landmarks[25];
    const left_ankle = landmarks[27];
    const right_hip = landmarks[24];
    const right_knee = landmarks[26];
    const right_ankle = landmarks[28];

    // Check if knees are aligned with ankles (side view)
    const left_knee_aligned = Math.abs(left_knee.x - left_ankle.x) < 0.1;
    const right_knee_aligned = Math.abs(right_knee.x - right_ankle.x) < 0.1;

    // Check if hips are level
    const hips_level = Math.abs(left_hip.y - right_hip.y) < 0.1;

    return left_knee_aligned && right_knee_aligned && hips_level;
  }

  detect_squat(landmarks) {
    // Get coordinates for left leg
    const left_hip = landmarks[23];
    const left_knee = landmarks[25];
    const left_ankle = landmarks[27];
    const left_shoulder = landmarks[11];

    // Get coordinates for right leg
    const right_hip = landmarks[24];
    const right_knee = landmarks[26];
    const right_ankle = landmarks[28];
    const right_shoulder = landmarks[12];

    // Calculate angles for left leg
    const left_knee_angle = this.calculate_angle(left_hip, left_knee, left_ankle);
    const left_hip_angle = this.calculate_angle(left_shoulder, left_hip, left_knee);

    // Calculate angles for right leg
    const right_knee_angle = this.calculate_angle(right_hip, right_knee, right_ankle);
    const right_hip_angle = this.calculate_angle(right_shoulder, right_hip, right_knee);

    // Average the angles
    const knee_angle = this.smooth_angle((left_knee_angle + right_knee_angle) / 2);
    const hip_angle = this.smooth_angle((left_hip_angle + right_hip_angle) / 2);

    // Validate form
    const form_valid = this.validate_form(landmarks);

    // Check if angles are within valid ranges
    if (knee_angle < 60 || knee_angle > 170 || hip_angle < 60 || hip_angle > 170) {
      return false;
    }

    // Check if both knee and hip angles are below threshold and form is valid
    return knee_angle < 110 && hip_angle < 140 && form_valid;
  }

  process_frame(landmarks) {
    const current_time = Date.now();
    const is_squat = this.detect_squat(landmarks);

    // Update consecutive frames counter
    if (is_squat) {
      this.consecutive_frames++;
    } else {
      this.consecutive_frames = 0;
    }

    // Check if we have enough consecutive frames and cooldown has passed
    if (this.consecutive_frames >= this.frames_required && 
        current_time - this.last_squat_time > this.cooldown) {
      if (this.stage === 'UP') {
        this.counter++;
        this.last_squat_time = current_time;
      }
      this.stage = 'DOWN';
    } else if (!is_squat) {
      this.stage = 'UP';
    }

    // Calculate angles for display
    const left_hip = landmarks[23];
    const left_knee = landmarks[25];
    const left_ankle = landmarks[27];
    const left_shoulder = landmarks[11];
    const right_hip = landmarks[24];
    const right_knee = landmarks[26];
    const right_ankle = landmarks[28];
    const right_shoulder = landmarks[12];

    const left_knee_angle = this.calculate_angle(left_hip, left_knee, left_ankle);
    const left_hip_angle = this.calculate_angle(left_shoulder, left_hip, left_knee);
    const right_knee_angle = this.calculate_angle(right_hip, right_knee, right_ankle);
    const right_hip_angle = this.calculate_angle(right_shoulder, right_hip, right_knee);

    return {
      counter: this.counter,
      stage: this.stage,
      angles: {
        left_knee: left_knee_angle,
        left_hip: left_hip_angle,
        right_knee: right_knee_angle,
        right_hip: right_hip_angle
      }
    };
  }

  reset() {
    this.counter = 0;
    this.stage = null;
    this.min_ang = 0;
    this.min_ang_hip = 0;
    this.last_squat_time = 0;
    this.consecutive_frames = 0;
    this.angle_history = [];
  }
}

module.exports = { SquatDetector }; 

File: backend/fitness/exercises/squats.py
import cv2
import mediapipe as mp
import numpy as np
import time

class SquatDetector:
    def __init__(self):
        self.mp_pose = mp.solutions.pose
        self.pose = self.mp_pose.Pose(
            static_image_mode=False,
            model_complexity=2,
            min_detection_confidence=0.7,
            min_tracking_confidence=0.7
        )
        self.mp_drawing = mp.solutions.drawing_utils
        self.mp_drawing_styles = mp.solutions.drawing_styles
        self.counter = 0
        self.stage = None
        self.min_ang = 0
        self.min_ang_hip = 0
        self.last_squat_time = 0
        self.squat_cooldown = 1.0
        self.consecutive_frames = 0
        self.required_frames = 5  # Further reduced for more responsive counting
        self.hip_angles = []
        self.knee_angles = []
        self.angle_window = 3  # Reduced window for faster response

    def calculate_angle(self, a, b, c):
        """Calculate angle between three points"""
        a = np.array(a)
        b = np.array(b)
        c = np.array(c)
        
        radians = np.arctan2(c[1]-b[1], c[0]-b[0]) - np.arctan2(a[1]-b[1], a[0]-b[0])
        angle = np.abs(radians*180.0/np.pi)
        
        if angle > 180.0:
            angle = 360-angle
        return angle

    def smooth_angle(self, angles, new_angle):
        """Apply moving average to smooth angle measurements"""
        angles.append(new_angle)
        if len(angles) > self.angle_window:
            angles.pop(0)
        return sum(angles) / len(angles)

    def detect_squat(self, landmarks):
        """Enhanced squat detection with form validation"""
        try:
            # Convert landmarks to the correct format
            landmarks_dict = {i: landmark for i, landmark in enumerate(landmarks)}

            # Get key points
            left_hip = [landmarks_dict[23]['x'], landmarks_dict[23]['y']]
            left_knee = [landmarks_dict[25]['x'], landmarks_dict[25]['y']]
            left_ankle = [landmarks_dict[27]['x'], landmarks_dict[27]['y']]
            left_shoulder = [landmarks_dict[11]['x'], landmarks_dict[11]['y']]

            right_hip = [landmarks_dict[24]['x'], landmarks_dict[24]['y']]
            right_knee = [landmarks_dict[26]['x'], landmarks_dict[26]['y']]
            right_ankle = [landmarks_dict[28]['x'], landmarks_dict[28]['y']]
            right_shoulder = [landmarks_dict[12]['x'], landmarks_dict[12]['y']]

            # Calculate angles
            left_knee_angle = self.calculate_angle(left_hip, left_knee, left_ankle)
            right_knee_angle = self.calculate_angle(right_hip, right_knee, right_ankle)
            left_hip_angle = self.calculate_angle(left_shoulder, left_hip, left_knee)
            right_hip_angle = self.calculate_angle(right_shoulder, right_hip, right_knee)

            # Smooth angles
            left_knee_angle = self.smooth_angle(self.knee_angles, left_knee_angle)
            right_knee_angle = self.smooth_angle(self.knee_angles, right_knee_angle)
            left_hip_angle = self.smooth_angle(self.hip_angles, left_hip_angle)
            right_hip_angle = self.smooth_angle(self.hip_angles, right_hip_angle)

            # Form validation
            knee_threshold = 110  # Slightly increased for easier detection
            hip_threshold = 140   # Slightly increased for easier detection
            min_knee_angle = 60   # Allow deeper squats
            
            # Check if angles are within valid ranges
            if not all(0 <= angle <= 180 for angle in [left_knee_angle, right_knee_angle, left_hip_angle, right_hip_angle]):
                return False, None

            # Check for proper squat form
            knees_bent = (left_knee_angle < knee_threshold and right_knee_angle < knee_threshold)
            hips_proper = (left_hip_angle < hip_threshold and right_hip_angle < hip_threshold)
            not_too_deep = (left_knee_angle > min_knee_angle and right_knee_angle > min_knee_angle)
            
            # Check for knee alignment (knees shouldn't go too far forward)
            knee_alignment = (abs(left_knee[0] - left_ankle[0]) < 0.15 and 
                            abs(right_knee[0] - right_ankle[0]) < 0.15)  # Slightly relaxed alignment check

            is_squat = knees_bent and hips_proper and not_too_deep and knee_alignment

            return is_squat, {
                'left_knee': left_knee_angle,
                'right_knee': right_knee_angle,
                'left_hip': left_hip_angle,
                'right_hip': right_hip_angle
            }

        except Exception as e:
            print(f"Error in squat detection: {str(e)}")
            return False, None

    def process_frame(self, landmarks):
        """Process a single frame for squat detection"""
        try:
            # Detect squat and get angles
            is_squat, angles = self.detect_squat(landmarks)

            if angles:
                # Update consecutive frames counter
                if is_squat:
                    self.consecutive_frames += 1
                else:
                    self.consecutive_frames = 0

                # Update counter and stage with improved logic
                current_time = time.time()
                
                if (self.consecutive_frames >= self.required_frames and 
                    self.stage != "DOWN" and 
                    current_time - self.last_squat_time > self.squat_cooldown):
                    self.stage = "DOWN"
                    self.counter += 1
                    self.min_ang = min(angles['left_knee'], angles['right_knee'])
                    self.min_ang_hip = min(angles['left_hip'], angles['right_hip'])
                    self.last_squat_time = current_time
                elif not is_squat and self.consecutive_frames == 0:
                    self.stage = "UP"

            return {
                "counter": self.counter,
                "stage": self.stage,
                "angles": angles if angles else None
            }

        except Exception as e:
            print(f"Error processing frame: {str(e)}")
            return {
                "counter": self.counter,
                "stage": self.stage,
                "angles": None
            }

    def reset(self):
        """Reset all counters and state"""
        self.counter = 0
        self.stage = None
        self.min_ang = 0
        self.min_ang_hip = 0
        self.last_squat_time = 0
        self.consecutive_frames = 0
        self.hip_angles = []
        self.knee_angles = [] 